<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad.utils.annotations &mdash; gnomad master documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="gnomad.utils.constraint" href="constraint.html" />
    <link rel="prev" title="gnomad.utils" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            gnomad
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/index.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../variant_qc/index.html">variant_qc</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">utils</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">annotations</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.pop_max_expr"><code class="docutils literal notranslate"><span class="pre">pop_max_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.project_max_expr"><code class="docutils literal notranslate"><span class="pre">project_max_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.faf_expr"><code class="docutils literal notranslate"><span class="pre">faf_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.gen_anc_faf_max_expr"><code class="docutils literal notranslate"><span class="pre">gen_anc_faf_max_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.qual_hist_expr"><code class="docutils literal notranslate"><span class="pre">qual_hist_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.age_hists_expr"><code class="docutils literal notranslate"><span class="pre">age_hists_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.get_lowqual_expr"><code class="docutils literal notranslate"><span class="pre">get_lowqual_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.get_annotations_hists"><code class="docutils literal notranslate"><span class="pre">get_annotations_hists()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.create_frequency_bins_expr"><code class="docutils literal notranslate"><span class="pre">create_frequency_bins_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.get_adj_expr"><code class="docutils literal notranslate"><span class="pre">get_adj_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.annotate_adj"><code class="docutils literal notranslate"><span class="pre">annotate_adj()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.add_variant_type"><code class="docutils literal notranslate"><span class="pre">add_variant_type()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.annotate_allele_info"><code class="docutils literal notranslate"><span class="pre">annotate_allele_info()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.annotation_type_is_numeric"><code class="docutils literal notranslate"><span class="pre">annotation_type_is_numeric()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.annotation_type_in_vcf_info"><code class="docutils literal notranslate"><span class="pre">annotation_type_in_vcf_info()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.bi_allelic_site_inbreeding_expr"><code class="docutils literal notranslate"><span class="pre">bi_allelic_site_inbreeding_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.fs_from_sb"><code class="docutils literal notranslate"><span class="pre">fs_from_sb()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.sor_from_sb"><code class="docutils literal notranslate"><span class="pre">sor_from_sb()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.pab_max_expr"><code class="docutils literal notranslate"><span class="pre">pab_max_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.bi_allelic_expr"><code class="docutils literal notranslate"><span class="pre">bi_allelic_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.unphase_call_expr"><code class="docutils literal notranslate"><span class="pre">unphase_call_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.region_flag_expr"><code class="docutils literal notranslate"><span class="pre">region_flag_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.missing_callstats_expr"><code class="docutils literal notranslate"><span class="pre">missing_callstats_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.set_female_y_metrics_to_na_expr"><code class="docutils literal notranslate"><span class="pre">set_female_y_metrics_to_na_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.hemi_expr"><code class="docutils literal notranslate"><span class="pre">hemi_expr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.merge_freq_arrays"><code class="docutils literal notranslate"><span class="pre">merge_freq_arrays()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.merge_histograms"><code class="docutils literal notranslate"><span class="pre">merge_histograms()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.annotate_freq"><code class="docutils literal notranslate"><span class="pre">annotate_freq()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.annotate_downsamplings"><code class="docutils literal notranslate"><span class="pre">annotate_downsamplings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.build_freq_stratification_list"><code class="docutils literal notranslate"><span class="pre">build_freq_stratification_list()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.generate_freq_group_membership_array"><code class="docutils literal notranslate"><span class="pre">generate_freq_group_membership_array()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.compute_freq_by_strata"><code class="docutils literal notranslate"><span class="pre">compute_freq_by_strata()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.update_structured_annotations"><code class="docutils literal notranslate"><span class="pre">update_structured_annotations()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.gks_compute_seqloc_digest"><code class="docutils literal notranslate"><span class="pre">gks_compute_seqloc_digest()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.add_gks_vrs"><code class="docutils literal notranslate"><span class="pre">add_gks_vrs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gnomad.utils.annotations.add_gks_va"><code class="docutils literal notranslate"><span class="pre">add_gks_va()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="constraint.html">constraint</a></li>
<li class="toctree-l3"><a class="reference internal" href="file_utils.html">file_utils</a></li>
<li class="toctree-l3"><a class="reference internal" href="filtering.html">filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="gen_stats.html">gen_stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="intervals.html">intervals</a></li>
<li class="toctree-l3"><a class="reference internal" href="liftover.html">liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="reference_genome.html">reference_genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="release.html">release</a></li>
<li class="toctree-l3"><a class="reference internal" href="slack.html">slack</a></li>
<li class="toctree-l3"><a class="reference internal" href="sparse_mt.html">sparse_mt</a></li>
<li class="toctree-l3"><a class="reference internal" href="vcf.html">vcf</a></li>
<li class="toctree-l3"><a class="reference internal" href="vep.html">vep</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../sample_qc/index.html">sample_qc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l2"><a class="reference internal" href="../assessment/index.html">assessment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../resource_sources.html">Resource Sources</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_methods/releases">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gnomad</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">gnomad</a></li>
          <li class="breadcrumb-item"><a href="index.html">gnomad.utils</a></li>
      <li class="breadcrumb-item active">gnomad.utils.annotations</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/api_reference/utils/annotations.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-utils-annotations">
<h1>gnomad.utils.annotations<a class="headerlink" href="#gnomad-utils-annotations" title="Permalink to this heading"></a></h1>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.pop_max_expr" title="gnomad.utils.annotations.pop_max_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.pop_max_expr</span></code></a>(freq, ...)</p></td>
<td><p>Create an expression containing the frequency information about the population that has the highest AF in <cite>freq_meta</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.project_max_expr" title="gnomad.utils.annotations.project_max_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.project_max_expr</span></code></a>(...)</p></td>
<td><p>Create an expression that computes allele frequency information by project for the <cite>n_projects</cite> with the largest AF at this row.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.faf_expr" title="gnomad.utils.annotations.faf_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.faf_expr</span></code></a>(freq, ...)</p></td>
<td><p>Calculate the filtering allele frequency (FAF) for each threshold specified in <cite>faf_thresholds</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.gen_anc_faf_max_expr" title="gnomad.utils.annotations.gen_anc_faf_max_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.gen_anc_faf_max_expr</span></code></a>(...)</p></td>
<td><p>Retrieve the maximum FAF and corresponding genetic ancestry for each of the thresholds in <cite>faf</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.qual_hist_expr" title="gnomad.utils.annotations.qual_hist_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.qual_hist_expr</span></code></a>([...])</p></td>
<td><p>Return a struct expression with genotype quality histograms based on the arguments given (dp, gq, ad, ab).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.age_hists_expr" title="gnomad.utils.annotations.age_hists_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.age_hists_expr</span></code></a>(...)</p></td>
<td><p>Return a StructExpression with the age histograms for hets and homs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.get_lowqual_expr" title="gnomad.utils.annotations.get_lowqual_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.get_lowqual_expr</span></code></a>(...)</p></td>
<td><p>Compute lowqual threshold expression for either split or unsplit alleles based on QUALapprox or AS_QUALapprox.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.get_annotations_hists" title="gnomad.utils.annotations.get_annotations_hists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.get_annotations_hists</span></code></a>(ht, ...)</p></td>
<td><p>Create histograms for variant metrics in ht.info.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.create_frequency_bins_expr" title="gnomad.utils.annotations.create_frequency_bins_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.create_frequency_bins_expr</span></code></a>(AC, AF)</p></td>
<td><p>Create bins for frequencies in preparation for aggregating QUAL by frequency bin.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.get_adj_expr" title="gnomad.utils.annotations.get_adj_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.get_adj_expr</span></code></a>(...[, ...])</p></td>
<td><p>Get adj genotype annotation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotate_adj" title="gnomad.utils.annotations.annotate_adj"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotate_adj</span></code></a>(mt[, ...])</p></td>
<td><p>Annotate genotypes with adj criteria (assumes diploid).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.add_variant_type" title="gnomad.utils.annotations.add_variant_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.add_variant_type</span></code></a>(...)</p></td>
<td><p>Get Struct of variant_type and n_alt_alleles from ArrayExpression of Strings.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotate_allele_info" title="gnomad.utils.annotations.annotate_allele_info"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotate_allele_info</span></code></a>(ht)</p></td>
<td><p>Return bi-allelic sites Table with an 'allele_info' annotation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotation_type_is_numeric" title="gnomad.utils.annotations.annotation_type_is_numeric"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotation_type_is_numeric</span></code></a>(t)</p></td>
<td><p>Given an annotation type, return whether it is a numerical type or not.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotation_type_in_vcf_info" title="gnomad.utils.annotations.annotation_type_in_vcf_info"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotation_type_in_vcf_info</span></code></a>(t)</p></td>
<td><p>Given an annotation type, returns whether that type can be natively exported to a VCF INFO field.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.bi_allelic_site_inbreeding_expr" title="gnomad.utils.annotations.bi_allelic_site_inbreeding_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.bi_allelic_site_inbreeding_expr</span></code></a>([...])</p></td>
<td><p>Return the site inbreeding coefficient as an expression to be computed on a MatrixTable.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.fs_from_sb" title="gnomad.utils.annotations.fs_from_sb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.fs_from_sb</span></code></a>(sb[, ...])</p></td>
<td><p>Compute <cite>FS</cite> (Fisher strand balance) annotation from  the <cite>SB</cite> (strand balance table) field.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.sor_from_sb" title="gnomad.utils.annotations.sor_from_sb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.sor_from_sb</span></code></a>(sb)</p></td>
<td><p>Compute <cite>SOR</cite> (Symmetric Odds Ratio test) annotation from  the <cite>SB</cite> (strand balance table) field.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.pab_max_expr" title="gnomad.utils.annotations.pab_max_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.pab_max_expr</span></code></a>(...[, ...])</p></td>
<td><p>Compute the maximum p-value of the binomial test for the alternate allele balance (PAB) for each allele.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.bi_allelic_expr" title="gnomad.utils.annotations.bi_allelic_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.bi_allelic_expr</span></code></a>(t)</p></td>
<td><p>Return a boolean expression selecting bi-allelic sites only, accounting for whether the input MT/HT was split.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.unphase_call_expr" title="gnomad.utils.annotations.unphase_call_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.unphase_call_expr</span></code></a>(...)</p></td>
<td><p>Generate unphased version of a call expression (which can be phased or not).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.region_flag_expr" title="gnomad.utils.annotations.region_flag_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.region_flag_expr</span></code></a>(t)</p></td>
<td><p>Create a <cite>region_flag</cite> struct that contains flags for problematic regions (i.e., LCR, decoy, segdup, and nonpar regions).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.missing_callstats_expr" title="gnomad.utils.annotations.missing_callstats_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.missing_callstats_expr</span></code></a>()</p></td>
<td><p>Create a missing callstats struct for insertion into frequency annotation arrays when data is missing.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.set_female_y_metrics_to_na_expr" title="gnomad.utils.annotations.set_female_y_metrics_to_na_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.set_female_y_metrics_to_na_expr</span></code></a>(t)</p></td>
<td><p>Set Y-variant frequency callstats for female-specific metrics to missing structs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.hemi_expr" title="gnomad.utils.annotations.hemi_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.hemi_expr</span></code></a>(locus, ...)</p></td>
<td><p>Return whether genotypes are hemizygous.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.merge_freq_arrays" title="gnomad.utils.annotations.merge_freq_arrays"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.merge_freq_arrays</span></code></a>(...)</p></td>
<td><p>Merge a list of frequency arrays based on the supplied <cite>operation</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.merge_histograms" title="gnomad.utils.annotations.merge_histograms"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.merge_histograms</span></code></a>(hists)</p></td>
<td><p>Merge a list of histogram annotations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotate_freq" title="gnomad.utils.annotations.annotate_freq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotate_freq</span></code></a>(mt[, ...])</p></td>
<td><p>Annotate <cite>mt</cite> with stratified allele frequencies.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotate_downsamplings" title="gnomad.utils.annotations.annotate_downsamplings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotate_downsamplings</span></code></a>(t, ...)</p></td>
<td><p>Annotate MatrixTable or Table with downsampling groups.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.build_freq_stratification_list" title="gnomad.utils.annotations.build_freq_stratification_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.build_freq_stratification_list</span></code></a>([...])</p></td>
<td><p>Build a list of stratification groupings to be used in frequency calculations based on supplied parameters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.generate_freq_group_membership_array" title="gnomad.utils.annotations.generate_freq_group_membership_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.generate_freq_group_membership_array</span></code></a>(ht, ...)</p></td>
<td><p>Generate a Table with a 'group_membership' array for each sample indicating whether the sample belongs to specific stratification groups.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.compute_freq_by_strata" title="gnomad.utils.annotations.compute_freq_by_strata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.compute_freq_by_strata</span></code></a>(mt)</p></td>
<td><p>Compute call statistics and, when passed, entry aggregation function(s) by strata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.update_structured_annotations" title="gnomad.utils.annotations.update_structured_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.update_structured_annotations</span></code></a>(ht, ...)</p></td>
<td><p>Update highly structured annotations on a Table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.gks_compute_seqloc_digest" title="gnomad.utils.annotations.gks_compute_seqloc_digest"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.gks_compute_seqloc_digest</span></code></a>(ht)</p></td>
<td><p>Compute sequence location digest-based id for a hail variant Table.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.add_gks_vrs" title="gnomad.utils.annotations.add_gks_vrs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.add_gks_vrs</span></code></a>(...)</p></td>
<td><p>Generate a dictionary containing VRS information from a given locus and struct of VRS information.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.add_gks_va" title="gnomad.utils.annotations.add_gks_va"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.add_gks_va</span></code></a>(input_struct)</p></td>
<td><p>Generate Python dictionary containing GKS VA annotations.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.utils.annotations"></span><dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.pop_max_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">pop_max_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">freq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_meta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pops_to_exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pop'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#pop_max_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.pop_max_expr" title="Permalink to this definition"></a></dt>
<dd><p>Create an expression containing the frequency information about the population that has the highest AF in <cite>freq_meta</cite>.</p>
<p>Populations specified in <cite>pops_to_exclude</cite> are excluded and only frequencies from adj populations are considered.</p>
<p>This resulting struct contains the following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>AC: int32</p></li>
<li><p>AF: float64</p></li>
<li><p>AN: int32</p></li>
<li><p>homozygote_count: int32</p></li>
<li><p>pop: str</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>freq</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of Structs with fields [‘AC’, ‘AF’, ‘AN’, ‘homozygote_count’]</p></li>
<li><p><strong>freq_meta</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of meta dictionaries corresponding to freq (as returned by annotate_freq)</p></li>
<li><p><strong>pops_to_exclude</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Set of populations to skip for popmax calcluation</p></li>
<li><p><strong>pop_label</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Label of the population field in the meta dictionary</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Popmax struct</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.project_max_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">project_max_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">project_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gt_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_projects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#project_max_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.project_max_expr" title="Permalink to this definition"></a></dt>
<dd><p>Create an expression that computes allele frequency information by project for the <cite>n_projects</cite> with the largest AF at this row.</p>
<p>Will return an array with one element per non-reference allele.</p>
<p>Each of these elements is itself an array of structs with the following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>AC: int32</p></li>
<li><p>AF: float64</p></li>
<li><p>AN: int32</p></li>
<li><p>homozygote_count: int32</p></li>
<li><p>project: str</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only projects with AF &gt; 0 are returned.
In case of ties, the project ordering is not guaranteed, and at most <cite>n_projects</cite> are returned.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>project_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>) – column expression containing the project</p></li>
<li><p><strong>gt_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – entry expression containing the genotype</p></li>
<li><p><strong>alleles_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – row expression containing the alleles</p></li>
<li><p><strong>n_projects</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of projects to return for each row</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>projectmax expression</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.faf_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">faf_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">freq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_meta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pops_to_exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">faf_thresholds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0.95,</span> <span class="pre">0.99]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pop'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#faf_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.faf_expr" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the filtering allele frequency (FAF) for each threshold specified in <cite>faf_thresholds</cite>.</p>
<p>See <a class="reference external" href="http://cardiodb.org/allelefrequencyapp/">http://cardiodb.org/allelefrequencyapp/</a> for more information.</p>
<p>The FAF is computed for each of the following population stratification if found in <cite>freq_meta</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><p>All samples, with adj criteria</p></li>
<li><p>For each population, with adj criteria</p></li>
<li><p>For all sex/population on the non-PAR regions of sex chromosomes (will be missing on autosomes and PAR regions of sex chromosomes)</p></li>
</ul>
</div></blockquote>
<p>Each of the FAF entry is a struct with one entry per threshold specified in <cite>faf_thresholds</cite> of type float64.</p>
<p>This returns a tuple with two expressions:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>An array of FAF expressions as described above</p></li>
<li><p>An array of dict containing the metadata for each of the array elements, in the same format as that produced by <cite>annotate_freq</cite>.</p></li>
</ol>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>freq</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of call stats structs (typically generated by hl.agg.call_stats)</p></li>
<li><p><strong>freq_meta</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of meta dictionaries corresponding to freq (typically generated using annotate_freq)</p></li>
<li><p><strong>locus</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.LocusExpression.html#hail.expr.LocusExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">LocusExpression</span></code></a>) – locus</p></li>
<li><p><strong>pops_to_exclude</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Set of populations to exclude from faf calculation (typically bottlenecked or consanguineous populations)</p></li>
<li><p><strong>faf_thresholds</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – List of FAF thresholds to compute</p></li>
<li><p><strong>pop_label</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Label of the population field in the meta dictionary</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>(FAF expression, FAF metadata)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.gen_anc_faf_max_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">gen_anc_faf_max_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">faf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">faf_meta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pop'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#gen_anc_faf_max_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.gen_anc_faf_max_expr" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve the maximum FAF and corresponding genetic ancestry for each of the thresholds in <cite>faf</cite>.</p>
<p>This resulting struct contains the following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>faf95_max: float64</p></li>
<li><p>faf95_max_gen_anc: str</p></li>
<li><p>faf99_max: float64</p></li>
<li><p>faf99_max_gen_anc: str</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>faf</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of Structs of FAF thresholds previously computed. When
<cite>faf_expr</cite> is used, contains fields ‘faf95’ and ‘faf99’.</p></li>
<li><p><strong>faf_meta</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of meta dictionaries corresponding to faf (as
returned by faf_expr)</p></li>
<li><p><strong>pop_label</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Label of the population field in the meta dictionary</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Genetic ancestry group struct for FAF max</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.qual_hist_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">qual_hist_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gt_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gq_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dp_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ad_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ab_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_adj_and_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#qual_hist_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.qual_hist_expr" title="Permalink to this definition"></a></dt>
<dd><p>Return a struct expression with genotype quality histograms based on the arguments given (dp, gq, ad, ab).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>If <cite>gt_expr</cite> is provided, will return histograms for non-reference samples only as well as all samples.</p></li>
<li><p><cite>gt_expr</cite> is required for the allele-balance histogram, as it is only computed on het samples.</p></li>
<li><p>If <cite>ab_expr</cite> is provided, the allele-balance histogram is computed using this expression instead of the ad_expr.</p></li>
<li><p>If <cite>adj_expr</cite> is provided, additional histograms are computed using only adj samples.</p></li>
</ul>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gt_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>]) – Entry expression containing genotype.</p></li>
<li><p><strong>gq_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – Entry expression containing genotype quality.</p></li>
<li><p><strong>dp_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – Entry expression containing depth.</p></li>
<li><p><strong>ad_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>]) – Entry expression containing allelic depth (bi-allelic here).</p></li>
<li><p><strong>adj_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Entry expression containing adj (high quality) genotype status.</p></li>
<li><p><strong>ab_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – Entry expression containing allele balance (bi-allelic here).</p></li>
<li><p><strong>split_adj_and_raw</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to split the adj and raw histograms into separate fields in the returned struct expr.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Genotype quality histograms expression.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.age_hists_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">age_hists_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gt_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">age_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowest_boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highest_boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#age_hists_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.age_hists_expr" title="Permalink to this definition"></a></dt>
<dd><p>Return a StructExpression with the age histograms for hets and homs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adj_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>) – Entry expression containing whether a genotype is high quality (adj) or not</p></li>
<li><p><strong>gt_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – Entry expression containing the genotype</p></li>
<li><p><strong>age_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Col expression containing the sample’s age</p></li>
<li><p><strong>lowest_boundary</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Lowest bin boundary (any younger sample will be binned in n_smaller)</p></li>
<li><p><strong>highest_boundary</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Highest bin boundary (any older sample will be binned in n_larger)</p></li>
<li><p><strong>n_bins</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Total number of bins</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A struct with <cite>age_hist_het</cite> and <cite>age_hist_hom</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.get_lowqual_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">get_lowqual_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qual_approx_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snv_phred_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snv_phred_het_prior</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indel_phred_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indel_phred_het_prior</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">39</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#get_lowqual_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.get_lowqual_expr" title="Permalink to this definition"></a></dt>
<dd><p>Compute lowqual threshold expression for either split or unsplit alleles based on QUALapprox or AS_QUALapprox.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When running This lowqual annotation using QUALapprox, it differs from the GATK LowQual filter.
This is because GATK computes this annotation at the site level, which uses the least stringent prior for mixed sites.
When run using AS_QUALapprox, this implementation can thus be more stringent for certain alleles at mixed sites.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alleles</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – Array of alleles</p></li>
<li><p><strong>qual_approx_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – QUALapprox or AS_QUALapprox</p></li>
<li><p><strong>snv_phred_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled SNV “emission” threshold (similar to GATK emission threshold)</p></li>
<li><p><strong>snv_phred_het_prior</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled SNV heterozygosity prior (30 = 1/1000 bases, GATK default)</p></li>
<li><p><strong>indel_phred_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled indel “emission” threshold (similar to GATK emission threshold)</p></li>
<li><p><strong>indel_phred_het_prior</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled indel heterozygosity prior (30 = 1/1000 bases, GATK default)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>lowqual expression (BooleanExpression if <cite>qual_approx_expr`is Numeric, Array[BooleanExpression] if `qual_approx_expr</cite> is ArrayNumeric)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.get_annotations_hists">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">get_annotations_hists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations_hists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log10_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['DP']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#get_annotations_hists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.get_annotations_hists" title="Permalink to this definition"></a></dt>
<dd><p>Create histograms for variant metrics in ht.info.</p>
<p>Used when creating site quality distribution json files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with variant metrics</p></li>
<li><p><strong>annotations_hists</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>]) – Dictionary of metrics names and their histogram values (start, end, bins)</p></li>
<li><p><strong>log10_annotations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of metrics to log scale</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of merics and their histograms</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict[str, hl.expr.StructExpression]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.create_frequency_bins_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">create_frequency_bins_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">AC</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">AF</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#create_frequency_bins_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.create_frequency_bins_expr" title="Permalink to this definition"></a></dt>
<dd><p>Create bins for frequencies in preparation for aggregating QUAL by frequency bin.</p>
<dl class="simple">
<dt>Bins:</dt><dd><ul class="simple">
<li><p>singleton</p></li>
<li><p>doubleton</p></li>
<li><p>0.00005</p></li>
<li><p>0.0001</p></li>
<li><p>0.0002</p></li>
<li><p>0.0005</p></li>
<li><p>0.001,</p></li>
<li><p>0.002</p></li>
<li><p>0.005</p></li>
<li><p>0.01</p></li>
<li><p>0.02</p></li>
<li><p>0.05</p></li>
<li><p>0.1</p></li>
<li><p>0.2</p></li>
<li><p>0.5</p></li>
<li><p>1</p></li>
</ul>
</dd>
</dl>
<p>NOTE: Frequencies should be frequencies from raw data.
Used when creating site quality distribution json files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>AC</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Field in input that contains the allele count information</p></li>
<li><p><strong>AF</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Field in input that contains the allele frequency information</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Expression containing bin name</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>hl.expr.StringExpression</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.get_adj_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">get_adj_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gt_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gq_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dp_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ad_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_gq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_dp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_ab</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">haploid_adj_dp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#get_adj_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.get_adj_expr" title="Permalink to this definition"></a></dt>
<dd><p>Get adj genotype annotation.</p>
<p>Defaults correspond to gnomAD values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gt_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – </p></li>
<li><p><strong>gq_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int32Expression.html#hail.expr.Int32Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int32Expression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int64Expression.html#hail.expr.Int64Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int64Expression</span></code></a>]) – </p></li>
<li><p><strong>dp_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int32Expression.html#hail.expr.Int32Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int32Expression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int64Expression.html#hail.expr.Int64Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int64Expression</span></code></a>]) – </p></li>
<li><p><strong>ad_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>) – </p></li>
<li><p><strong>adj_gq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>adj_dp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>adj_ab</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>haploid_adj_dp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.annotate_adj">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">annotate_adj</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_gq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_dp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_ab</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">haploid_adj_dp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotate_adj"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.annotate_adj" title="Permalink to this definition"></a></dt>
<dd><p>Annotate genotypes with adj criteria (assumes diploid).</p>
<p>Defaults correspond to gnomAD values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – </p></li>
<li><p><strong>adj_gq</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>adj_dp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
<li><p><strong>adj_ab</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
<li><p><strong>haploid_adj_dp</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.add_variant_type">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">add_variant_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alt_alleles</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#add_variant_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.add_variant_type" title="Permalink to this definition"></a></dt>
<dd><p>Get Struct of variant_type and n_alt_alleles from ArrayExpression of Strings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>alt_alleles</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – </p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.annotate_allele_info">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">annotate_allele_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotate_allele_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.annotate_allele_info" title="Permalink to this definition"></a></dt>
<dd><p>Return bi-allelic sites Table with an ‘allele_info’ annotation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function requires that the input <cite>ht</cite> is unsplit and returns a split <cite>ht</cite>.</p>
</div>
<dl class="simple">
<dt>‘allele_info’ is a struct with the following information:</dt><dd><ul class="simple">
<li><p>variant_type: Variant type (snv, indel, multi-snv, multi-indel, or mixed).</p></li>
<li><p>n_alt_alleles: Total number of alternate alleles observed at variant locus.</p></li>
<li><p>has_star: True if the variant contains a star allele.</p></li>
<li><p>allele_type: Allele type (snv, insertion, deletion, or mixed).</p></li>
<li><p>was_mixed: True if the variant was mixed (i.e. contained both SNVs and indels).</p></li>
<li><p>nonsplit_alleles: Array of alleles before splitting.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Unsplit input Table.</p></li>
<li><p><strong>ht</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Split Table with allele data annotation added,</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.annotation_type_is_numeric">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">annotation_type_is_numeric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotation_type_is_numeric"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.annotation_type_is_numeric" title="Permalink to this definition"></a></dt>
<dd><p>Given an annotation type, return whether it is a numerical type or not.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code>) – Type to test</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>If the input type is numeric</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.annotation_type_in_vcf_info">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">annotation_type_in_vcf_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotation_type_in_vcf_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.annotation_type_in_vcf_info" title="Permalink to this definition"></a></dt>
<dd><p>Given an annotation type, returns whether that type can be natively exported to a VCF INFO field.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Types that aren’t natively exportable to VCF will be converted to String on export.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code>) – Type to test</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>If the input type can be exported to VCF</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.bi_allelic_site_inbreeding_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">bi_allelic_site_inbreeding_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">call</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callstats_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#bi_allelic_site_inbreeding_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.bi_allelic_site_inbreeding_expr" title="Permalink to this definition"></a></dt>
<dd><p>Return the site inbreeding coefficient as an expression to be computed on a MatrixTable.</p>
<p>This is implemented based on the GATK InbreedingCoeff metric:
<a class="reference external" href="https://software.broadinstitute.org/gatk/documentation/article.php?id=8032">https://software.broadinstitute.org/gatk/documentation/article.php?id=8032</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The computation is run based on the counts of alternate alleles and thus should only be run on bi-allelic sites.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>call</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>]) – Expression giving the calls in the MT</p></li>
<li><p><strong>callstats_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a>]) – StructExpression containing only alternate allele AC, AN, and homozygote_count as integers. If passed, used to create expression in place of GT calls.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Float32Expression.html#hail.expr.Float32Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Float32Expression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Site inbreeding coefficient expression</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.fs_from_sb">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">fs_from_sb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cell_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_p_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-320</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#fs_from_sb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.fs_from_sb" title="Permalink to this definition"></a></dt>
<dd><p>Compute <cite>FS</cite> (Fisher strand balance) annotation from  the <cite>SB</cite> (strand balance table) field.</p>
<p><cite>FS</cite> is the phred-scaled value of the double-sided Fisher exact test on strand balance.</p>
<p>Using default values will have the same behavior as the GATK implementation, that is:
- If sum(counts) &gt; 2*`min_cell_count` (default to GATK value of 200), they are normalized
- If sum(counts) &lt; <cite>min_count</cite> (default to GATK value of 4), returns missing
- Any p-value &lt; <cite>min_p_value</cite> (default to GATK value of 1e-320) is truncated to that value</p>
<p>In addition to the default GATK behavior, setting <cite>normalize</cite> to <cite>False</cite> will perform a chi-squared test
for large counts (&gt; <cite>min_cell_count</cite>) instead of normalizing the cell values.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function can either take
- an array of length four containing the forward and reverse strands’ counts of ref and alt alleles: [ref fwd, ref rev, alt fwd, alt rev]
- a two dimensional array with arrays of length two, containing the counts: [[ref fwd, ref rev], [alt fwd, alt rev]]</p>
</div>
<p>GATK code here: <a class="reference external" href="https://github.com/broadinstitute/gatk/blob/master/src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/FisherStrand.java">https://github.com/broadinstitute/gatk/blob/master/src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/FisherStrand.java</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sb</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]) – Count of ref/alt reads on each strand</p></li>
<li><p><strong>normalize</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to normalize counts is sum(counts) &gt; min_cell_count (normalize=True), or use a chi sq instead of FET (normalize=False)</p></li>
<li><p><strong>min_cell_count</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum count for performing a FET</p></li>
<li><p><strong>min_count</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Minimum total count to output FS (otherwise null it output)</p></li>
<li><p><strong>min_p_value</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int64Expression.html#hail.expr.Int64Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int64Expression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>FS value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.sor_from_sb">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">sor_from_sb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sb</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#sor_from_sb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.sor_from_sb" title="Permalink to this definition"></a></dt>
<dd><p>Compute <cite>SOR</cite> (Symmetric Odds Ratio test) annotation from  the <cite>SB</cite> (strand balance table) field.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function can either take
- an array of length four containing the forward and reverse strands’ counts of ref and alt alleles: [ref fwd, ref rev, alt fwd, alt rev]
- a two dimensional array with arrays of length two, containing the counts: [[ref fwd, ref rev], [alt fwd, alt rev]]</p>
</div>
<p>GATK code here: <a class="reference external" href="https://github.com/broadinstitute/gatk/blob/master/src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/StrandOddsRatio.java">https://github.com/broadinstitute/gatk/blob/master/src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/StrandOddsRatio.java</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>sb</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]) – Count of ref/alt reads on each strand</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Float64Expression.html#hail.expr.Float64Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Float64Expression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>SOR value</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.pab_max_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">pab_max_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gt_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ad_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">la_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_alleles_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#pab_max_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.pab_max_expr" title="Permalink to this definition"></a></dt>
<dd><p>Compute the maximum p-value of the binomial test for the alternate allele balance (PAB) for each allele.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function can take a <cite>gt_expr</cite> and <cite>ad_expr</cite> that use local or global
alleles. If they use local alleles, <cite>la_expr</cite> and <cite>n_alleles_expr</cite> should be
provided to transform <cite>gt_expr</cite> and <cite>ad_expr</cite> to global alleles.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gt_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – Genotype call expression.</p></li>
<li><p><strong>ad_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – Allele depth expression.</p></li>
<li><p><strong>la_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]) – Allele local index expression. When provided <cite>gt_expr</cite> and
<cite>ad_expr</cite> are transformed from using local alleles to global alleles using
<cite>la_expr</cite>.</p></li>
<li><p><strong>n_alleles_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int32Expression.html#hail.expr.Int32Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int32Expression</span></code></a>]) – Number of alleles expression. Required when ‘la_expr’ is
provided.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array expression of maximum p-values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.bi_allelic_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">bi_allelic_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#bi_allelic_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.bi_allelic_expr" title="Permalink to this definition"></a></dt>
<dd><p>Return a boolean expression selecting bi-allelic sites only, accounting for whether the input MT/HT was split.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]) – Input HT/MT</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Boolean expression selecting only bi-allelic sites</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.unphase_call_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">unphase_call_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">call_expr</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#unphase_call_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.unphase_call_expr" title="Permalink to this definition"></a></dt>
<dd><p>Generate unphased version of a call expression (which can be phased or not).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>call_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – Input call expression</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>unphased call expression</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.region_flag_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">region_flag_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_par</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prob_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#region_flag_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.region_flag_expr" title="Permalink to this definition"></a></dt>
<dd><p>Create a <cite>region_flag</cite> struct that contains flags for problematic regions (i.e., LCR, decoy, segdup, and nonpar regions).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>No hg38 resources for decoy or self chain are available yet.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]) – Input Table/MatrixTable</p></li>
<li><p><strong>non_par</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, flag loci that occur within pseudoautosomal regions on sex chromosomes</p></li>
<li><p><strong>prob_regions</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – If supplied, flag loci that occur within regions defined in Hail Table(s)</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>region_flag</cite> struct row annotation</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.missing_callstats_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">missing_callstats_expr</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#missing_callstats_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.missing_callstats_expr" title="Permalink to this definition"></a></dt>
<dd><p>Create a missing callstats struct for insertion into frequency annotation arrays when data is missing.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Hail Struct with missing values for each callstats element</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.set_female_y_metrics_to_na_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">set_female_y_metrics_to_na_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'freq'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_meta_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'freq_meta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_index_dict_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'freq_index_dict'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#set_female_y_metrics_to_na_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.set_female_y_metrics_to_na_expr" title="Permalink to this definition"></a></dt>
<dd><p>Set Y-variant frequency callstats for female-specific metrics to missing structs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]) – Table or MatrixTable for which to adjust female metrics.</p></li>
<li><p><strong>freq_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Array expression or string annotation name for the frequency
array. Default is “freq”.</p></li>
<li><p><strong>freq_meta_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Array expression or string annotation name for the frequency
metadata. Default is “freq_meta”.</p></li>
<li><p><strong>freq_index_dict_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.DictExpression.html#hail.expr.DictExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DictExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Dict expression or string annotation name for the
frequency metadata index dictionary. Default is “freq_index_dict”.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Hail array expression to set female Y-variant metrics to missing values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.hemi_expr">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">hemi_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sex_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">male_str</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'XY'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#hemi_expr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.hemi_expr" title="Permalink to this definition"></a></dt>
<dd><p>Return whether genotypes are hemizygous.</p>
<p>Return missing expression if locus is not in chrX/chrY non-PAR regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locus</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.LocusExpression.html#hail.expr.LocusExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">LocusExpression</span></code></a>) – Input locus.</p></li>
<li><p><strong>sex_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>) – Input StringExpression indicating whether sample is XX or XY.</p></li>
<li><p><strong>gt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – Input genotype.</p></li>
<li><p><strong>xy_str</strong> – String indicating whether sample is XY. Default is “XY”.</p></li>
<li><p><strong>male_str</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>BooleanExpression indicating whether genotypes are hemizygous.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.merge_freq_arrays">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">merge_freq_arrays</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">farrays</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fmeta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">operation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_negatives_to_zero</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_arrays</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#merge_freq_arrays"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.merge_freq_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Merge a list of frequency arrays based on the supplied <cite>operation</cite>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Arrays must be on the same Table.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Arrays do not have to contain the same groupings or order of groupings but
the array indices for a freq array in <cite>farrays</cite> must be the same as its associated
frequency metadata index in <cite>fmeta</cite> i.e., <cite>farrays = [freq1, freq2]</cite> then <cite>fmeta</cite>
must equal <cite>[fmeta1, fmeta2]</cite> where fmeta1 contains the metadata information
for freq1.</p>
<p>If <cite>operation</cite> is set to “sum”, groups in the merged array
will be the union of groupings found within the arrays’ metadata and all arrays
with be summed by grouping. If <cite>operation</cite> is set to “diff”, the merged array
will contain groups only found in the first array of <cite>fmeta</cite>. Any array containing
any of these groups will have thier values subtracted from the values of the first array.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>farrays</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]) – List of frequency arrays to merge. First entry in the list is the primary array to which other arrays will be added or subtracted. All arrays must be on the same Table.</p></li>
<li><p><strong>fmeta</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]]) – List of frequency metadata for arrays being merged.</p></li>
<li><p><strong>operation</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Merge operation to perform. Options are “sum” and “diff”. If “diff” is passed, the first freq array in the list will have the other arrays subtracted from it.</p></li>
<li><p><strong>set_negatives_to_zero</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, set negative array values to 0 for AC, AN, AF, and homozygote_count. If False, raise a ValueError. Default is False.</p></li>
<li><p><strong>count_arrays</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]]]) – Dictionary of Lists of arrays containing counts to merge using the passed operation. Must use the same group indexing as fmeta. Keys are the descriptor names, values are Lists of arrays to merge. Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]], <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]], <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]]]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple of merged frequency array, frequency metadata list and if <cite>count_arrays</cite> is not None, a dictionary of merged count arrays.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.merge_histograms">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">merge_histograms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hists</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#merge_histograms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.merge_histograms" title="Permalink to this definition"></a></dt>
<dd><p>Merge a list of histogram annotations.</p>
<p>This function merges a list of histogram annotations by summing the arrays
in an element-wise fashion. It keeps one ‘bin_edge’ annotation but merges the
‘bin_freq’, ‘n_smaller’, and ‘n_larger’ annotations by summing them.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Bin edges are assumed to be the same for all histograms.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>hists</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a>]) – List of histogram structs to merge.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Expression.html#hail.expr.Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Expression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Merged histogram struct.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.annotate_freq">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">annotate_freq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sex_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subpop_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_strata_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsamplings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsampling_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds_pop_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry_agg_funcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_mt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotate_freq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.annotate_freq" title="Permalink to this definition"></a></dt>
<dd><p>Annotate <cite>mt</cite> with stratified allele frequencies.</p>
<dl class="simple">
<dt>The output Matrix table will include:</dt><dd><ul class="simple">
<li><p>row annotation <cite>freq</cite> containing the stratified allele frequencies</p></li>
<li><p>global annotation <cite>freq_meta</cite> with metadata</p></li>
<li><p>global annotation <cite>freq_meta_sample_count</cite> with sample count information</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently this only supports bi-allelic sites.</p>
<dl class="simple">
<dt>The input <cite>mt</cite> needs to have the following entry fields:</dt><dd><ul class="simple">
<li><p>GT: a CallExpression containing the genotype</p></li>
<li><p>adj: a BooleanExpression containing whether the genotype is of high quality
or not.</p></li>
</ul>
</dd>
</dl>
<p>All expressions arguments need to be expression on the input <cite>mt</cite>.</p>
</div>
<p class="rubric"><cite>freq</cite> row annotation</p>
<p>The <cite>freq</cite> row annotation is an Array of Structs, with each Struct containing the
following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>AC: int32</p></li>
<li><p>AF: float64</p></li>
<li><p>AN: int32</p></li>
<li><p>homozygote_count: int32</p></li>
</ul>
</div></blockquote>
<p>Each element of the array corresponds to a stratification of the data, and the
metadata about these annotations is stored in the globals.</p>
<p class="rubric">Global <cite>freq_meta</cite> metadata annotation</p>
<p>The global annotation <cite>freq_meta</cite> is added to the input <cite>mt</cite>. It is a list of dict.
Each element of the list contains metadata on a frequency stratification and the
index in the list corresponds to the index of that frequency stratification in the
<cite>freq</cite> row annotation.</p>
<p class="rubric">Global <cite>freq_meta_sample_count</cite> annotation</p>
<p>The global annotation <cite>freq_meta_sample_count</cite> is added to the input <cite>mt</cite>. This is a
sample count per sample grouping defined in the <cite>freq_meta</cite> global annotation.</p>
<p class="rubric">The <cite>additional_strata_expr</cite> parameter</p>
<p>If the <cite>additional_strata_expr</cite> parameter is used, frequencies will be computed for
each of the strata dictionaries across all values. For example, if
<cite>additional_strata_expr</cite> is set to <cite>[{‘platform’: mt.platform},
{‘platform’:mt.platform, ‘pop’: mt.pop}, {‘age_bin’: mt.age_bin}]</cite>, then
frequencies will be computed for each of the values of <cite>mt.platform</cite>, each of the
combined values of <cite>mt.platform</cite> and <cite>mt.pop</cite>, and each of the values of
<cite>mt.age_bin</cite>.</p>
<p class="rubric">The <cite>downsamplings</cite> parameter</p>
<p>If the <cite>downsamplings</cite> parameter is used without the <cite>downsampling_expr</cite>,
frequencies will be computed for all samples and by population (if <cite>pop_expr</cite> is
specified) by downsampling the number of samples without replacement to each of the
numbers specified in the <cite>downsamplings</cite> array, provided that there are enough
samples in the dataset. In addition, if <cite>pop_expr</cite> is specified, a downsampling to
each of the exact number of samples present in each population is added. Note that
samples are randomly sampled only once, meaning that the lower downsamplings are
subsets of the higher ones. If the <cite>downsampling_expr</cite> parameter is used with the
<cite>downsamplings</cite> parameter, the <cite>downsamplings</cite> parameter informs the function which
downsampling groups were already created and are to be used in the frequency
calculation.</p>
<p class="rubric">The <cite>downsampling_expr</cite> and <cite>ds_pop_counts</cite> parameters</p>
<p>If the <cite>downsampling_expr</cite> parameter is used, <cite>downsamplings</cite> must also be set
and frequencies will be computed for all samples and by population (if <cite>pop_expr</cite>
is specified) using the downsampling indices to each of the numbers specified in
the <cite>downsamplings</cite> array. The function expects a ‘global_idx’, and if <cite>pop_expr</cite>
is used, a ‘pop_idx’ within the <cite>downsampling_expr</cite> to be used to determine if a
sample belongs within a certain downsampling group, i.e. the index is less than
the group size. <cite>The function `annotate_downsamplings</cite> can be used to to create
the <cite>downsampling_expr</cite>, <cite>downsamplings</cite>, and <cite>ds_pop_counts</cite> expressions.</p>
<p class="rubric">The <cite>entry_agg_funcs</cite> parameter</p>
<p>If the <cite>entry_agg_funcs</cite> parameter is used, the output MatrixTable will also
contain the annotations specified in the <cite>entry_agg_funcs</cite> parameter. The keys of
the dict are the names of the annotations and the values are tuples of functions.
The first function is used to transform the <cite>mt</cite> entries in some way, and the
second function is used to aggregate the output from the first function. For
example, if <cite>entry_agg_funcs</cite> is set to {‘adj_samples’: (get_adj_expr, hl.agg.sum)}`,
then the output MatrixTable will contain an annotation <cite>adj_samples</cite> which is an
array of the number of adj samples per strata in each row.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable</p></li>
<li><p><strong>sex_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, frequencies are stratified by sex. If <cite>pop_expr</cite>
is also specified, then a pop/sex stratifiction is added.</p></li>
<li><p><strong>pop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, frequencies are stratified by population. If
<cite>sex_expr</cite> is also specified, then a pop/sex stratifiction is added.</p></li>
<li><p><strong>subpop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, frequencies are stratified by sub-continental
population. Note that <cite>pop_expr</cite> is required as well when using this option.</p></li>
<li><p><strong>additional_strata_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]], <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>], <code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code>]) – When specified, frequencies are stratified by the
given additional strata. This can e.g. be used to stratify by platform,
platform-pop, platform-pop-sex.</p></li>
<li><p><strong>downsamplings</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]) – When specified, frequencies are computed by downsampling the
data to the number of samples given in the list. Note that if <cite>pop_expr</cite> is
specified, downsamplings by population is also computed.</p></li>
<li><p><strong>downsampling_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a>]) – When specified, frequencies are computed using the
downsampling indices in the provided StructExpression. Note that if <cite>pop_idx</cite>
is specified within the struct, downsamplings by population is also computed.</p></li>
<li><p><strong>ds_pop_counts</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]) – When specified, frequencies are computed by downsampling the
data to the number of samples per pop in the dict. The key is the population
and the value is the number of samples.</p></li>
<li><p><strong>entry_agg_funcs</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Callable</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Callable</span></code>]]]) – When specified, additional annotations are added to the
output Table/MatrixTable. The keys of the dict are the names of the annotations
and the values are tuples of functions. The first function is used to transform
the <cite>mt</cite> entries in some way, and the second function is used to aggregate the
output from the first function.</p></li>
<li><p><strong>annotate_mt</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to return the full MatrixTable with annotations added
instead of only a Table with <cite>freq</cite> and other annotations. Default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>MatrixTable or Table with <cite>freq</cite> annotation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.annotate_downsamplings">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">annotate_downsamplings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsamplings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotate_downsamplings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.annotate_downsamplings" title="Permalink to this definition"></a></dt>
<dd><p>Annotate MatrixTable or Table with downsampling groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Input MatrixTable or Table.</p></li>
<li><p><strong>downsamplings</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – List of downsampling sizes.</p></li>
<li><p><strong>pop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – Optional expression for population group. When provided, population
sample sizes are added as values to downsamplings.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>MatrixTable or Table with downsampling annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.build_freq_stratification_list">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">build_freq_stratification_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sex_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subpop_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_strata_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsampling_expr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#build_freq_stratification_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.build_freq_stratification_list" title="Permalink to this definition"></a></dt>
<dd><p>Build a list of stratification groupings to be used in frequency calculations based on supplied parameters.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is primarily used through <cite>annotate_freq</cite> but can be used
independently if desired. The returned list of stratifications can be passed to
<cite>generate_freq_group_membership_array</cite>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sex_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, the returned list contains a stratification for
sex. If <cite>pop_expr</cite> is also specified, then the returned list also contains a
pop/sex stratification.</p></li>
<li><p><strong>pop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, the returned list contains a stratification for
population. If <cite>sex_expr</cite> is also specified, then the returned list also
contains a pop/sex stratification.</p></li>
<li><p><strong>subpop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, the returned list contains a stratification for
sub-continental population. Note that <cite>pop_expr</cite> is required as well when using
this option.</p></li>
<li><p><strong>additional_strata_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]], <code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>], <code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code>]) – When specified, the returned list contains a
stratification for each of the additional strata. This can e.g. be used to
stratify by platform, platform-pop, platform-pop-sex.</p></li>
<li><p><strong>downsampling_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a>]) – When specified, the returned list contains a
stratification for downsampling. If <cite>pop_expr</cite> is also specified, then the
returned list also contains a downsampling/pop stratification.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of dictionaries specifying stratification groups where the keys of
each dictionary are strings and the values are corresponding expressions that
define the values to stratify frequency calculations by.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.generate_freq_group_membership_array">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">generate_freq_group_membership_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strata_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsamplings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds_pop_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_zero_sample_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">no_raw_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#generate_freq_group_membership_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.generate_freq_group_membership_array" title="Permalink to this definition"></a></dt>
<dd><p>Generate a Table with a ‘group_membership’ array for each sample indicating whether the sample belongs to specific stratification groups.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is primarily used through <cite>annotate_freq</cite> but can be used
independently if desired. Please see the <cite>annotate_freq</cite> function for more
complete documentation.</p>
</div>
<dl class="simple">
<dt>The following global annotations are added to the returned Table:</dt><dd><ul class="simple">
<li><p>freq_meta: Each element of the list contains metadata on a stratification
group.</p></li>
<li><p>freq_meta_sample_count: sample count per grouping defined in <cite>freq_meta</cite>.</p></li>
<li><p>If downsamplings or ds_pop_counts are specified, they are also added as
global annotations on the returned Table.</p></li>
</ul>
</dd>
</dl>
<p>Each sample is annotated with a ‘group_membership’ array indicating whether the
sample belongs to specific stratification groups. All possible value combinations
are determined for each stratification grouping in the <cite>strata_expr</cite> list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input Table that contains Expressions specified by <cite>strata_expr</cite>.</p></li>
<li><p><strong>strata_expr</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]]) – List of dictionaries specifying stratification groups where
the keys of each dictionary are strings and the values are corresponding
expressions that define the values to stratify frequency calculations by.</p></li>
<li><p><strong>downsamplings</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]) – List of downsampling values to include in the stratifications.</p></li>
<li><p><strong>ds_pop_counts</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]) – Dictionary of population counts for each downsampling value.</p></li>
<li><p><strong>remove_zero_sample_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to remove groups with a sample count of 0.
Default is False.</p></li>
<li><p><strong>no_raw_group</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to remove the raw group from the ‘group_membership’
annotation and the ‘freq_meta’ and ‘freq_meta_sample_count’ global annotations.
Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table with the ‘group_membership’ array annotation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.compute_freq_by_strata">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">compute_freq_by_strata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry_agg_funcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">select_fields</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_membership_includes_raw_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#compute_freq_by_strata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.compute_freq_by_strata" title="Permalink to this definition"></a></dt>
<dd><p>Compute call statistics and, when passed, entry aggregation function(s) by strata.</p>
<p>The computed call statistics are AC, AF, AN, and homozygote_count. The entry
aggregation functions are applied to the MatrixTable entries and aggregated. The
MatrixTable must contain a ‘group_membership’ annotation (like the one added by
<cite>generate_freq_group_membership_array</cite>) that is a list of bools to aggregate the
columns by.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is primarily used through <cite>annotate_freq</cite> but can be used
independently if desired. Please see the <cite>annotate_freq</cite> function for more
complete documentation.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable.</p></li>
<li><p><strong>entry_agg_funcs</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Callable</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Callable</span></code>]]]) – Optional dict of entry aggregation functions. When
specified, additional annotations are added to the output Table/MatrixTable.
The keys of the dict are the names of the annotations and the values are tuples
of functions. The first function is used to transform the <cite>mt</cite> entries in some
way, and the second function is used to aggregate the output from the first
function.</p></li>
<li><p><strong>select_fields</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Optional list of row fields from <cite>mt</cite> to keep on the output
Table.</p></li>
<li><p><strong>group_membership_includes_raw_group</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether the ‘group_membership’
annotation includes an entry for the ‘raw’ group, representing all samples. If
False, the ‘raw’ group is inserted as the second element in all added
annotations using the same ‘group_membership’, resulting
in array lengths of ‘group_membership’+1. If True, the second element of each
added annotation is still the ‘raw’ group, but the group membership is
determined by the values in the second element of ‘group_membership’, and the
output annotations will be the same length as ‘group_membership’. Default is
True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table or MatrixTable with allele frequencies by strata.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.update_structured_annotations">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">update_structured_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_update_exprs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_update_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#update_structured_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.update_structured_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Update highly structured annotations on a Table.</p>
<p>This function recursively updates annotations defined by <cite>annotation_update_exprs</cite>
and if <cite>annotation_update_label</cite> is supplied, it checks if the sample annotations
are different from the input and adds a flag to the Table, indicating which
annotations have been updated for each sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input Table with structured annotations to update.</p></li>
<li><p><strong>annotation_update_exprs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Expression.html#hail.expr.Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Expression</span></code></a>]) – Dictionary of annotations to update, structured as
they are structured on the input <cite>ht</cite>.</p></li>
<li><p><strong>annotation_update_label</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Optional string of the label to use for an
annotation indicating which annotations have been updated. Default is None, so
no annotation is added.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table with updated annotations and optionally a flag indicating which
annotations were changed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.gks_compute_seqloc_digest">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">gks_compute_seqloc_digest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">export_tmpfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">computed_tmpfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#gks_compute_seqloc_digest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.gks_compute_seqloc_digest" title="Permalink to this definition"></a></dt>
<dd><p>Compute sequence location digest-based id for a hail variant Table.</p>
<p>Exports table to tsv, computes SequenceLocation digests, reimports and replaces
the vrs_json field with the result. Input table must have a .vrs field, like the
one added by add_gks_vrs, that can be used to construct ga4gh.vrs models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – hail table with VRS annotation</p></li>
<li><p><strong>export_tmpfile</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Optional file path to export the table to.</p></li>
<li><p><strong>computed_tmpfile</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Optional file path to write the updated rows to,
which is then imported as a hail table</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a hail table with the VRS annotation updated with the new SequenceLocations</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.add_gks_vrs">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">add_gks_vrs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_locus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_vrs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#add_gks_vrs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.add_gks_vrs" title="Permalink to this definition"></a></dt>
<dd><p>Generate a dictionary containing VRS information from a given locus and struct of VRS information.</p>
<p>Dict will have GA4GH GKS VRS structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_locus</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">locus</span></code>) – Locus field from a struct (locus of result of running .collect() on a Hail table).</p></li>
<li><p><strong>input_vrs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">struct</span></code>) – VRS struct (such as from a ht.info.vrs field).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Python dictionary conforming to GA4GH GKS VRS structure.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad.utils.annotations.add_gks_va">
<span class="sig-prename descclassname"><span class="pre">gnomad.utils.annotations.</span></span><span class="sig-name descname"><span class="pre">add_gks_va</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_struct</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gnomAD'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'3.1.2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestry_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestry_groups_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by_sex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq_index_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#add_gks_va"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad.utils.annotations.add_gks_va" title="Permalink to this definition"></a></dt>
<dd><p>Generate Python dictionary containing GKS VA annotations.</p>
<p>Populate the dictionary with frequency information conforming to the GKS VA frequency schema.
If ancestry_groups or by_sex is provided, also include subcohort schemas for each cohort.
If input_struct has mean_depth, it is added to ancillaryResults.
This annotation is added under the gks_va_freq_dict field of the table.
The focusAllele field is not populated, and must be filled in by the caller.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_struct</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">struct</span></code>) – Hail struct for a desired variant (such as result of running .collect()[0] on a Table).</p></li>
<li><p><strong>label_name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Label name to use within the returned dictionary. Example: “gnomAD”.</p></li>
<li><p><strong>label_version</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – String listing the version of the table being used. Example: “3.1.2”.</p></li>
<li><p><strong>ancestry_groups</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>) – List of strings of shortened names of cohorts to return results for.
Example: [‘afr’,’fin’,’nfe’]. Default is None.</p></li>
<li><p><strong>ancestry_groups_dict</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>) – Dict mapping shortened genetic ancestry group names to full names.
Example: {‘afr’:’African/African American’}. Default is None.</p></li>
<li><p><strong>by_sex</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Boolean to include breakdown of cohorts by inferred sex (XX and XY) as well.
Default is None.</p></li>
<li><p><strong>freq_index_dict</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Freq_index_dict<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict mapping groups to their index for freq info in ht.freq_index_dict[0].
Default is None.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple containing a dictionary containing GKS VA frequency information,
(split by ancestry groups and sex if desired) for the specified variant.</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="gnomad.utils" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="constraint.html" class="btn btn-neutral float-right" title="gnomad.utils.constraint" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>